### PR TITLE
Run INLA in 32-bit floating point

### PR DESCRIPTION
**Pull request by constinit**
We make a few changes to improve numerical stability:
- Using the faster_inverse routine for inverting the Hessian, which requires
only a single sherman-morrison application and is numerically stable
- Calculating the multivariate normal pdf using the eigendecomposition of
the covariance matrix, which is stabler and faster than the quadratic
form
- Using the stabler slogdet numpy function

Combined, we reduce runtime by *50%* while retaining
accuracy within 0.1%.

We still do the initial setup in full 64-bit precision.

We also cap the maximum number of Newton iterations to 100 to prevent infinite
loops.

### CODE REVIEW

File: research/berry/berrylib/fast_inla.py [hunk 1]
```diff
@@ -328,59 +337,94 @@ def step(args):
...
+        "logit_p1",
+        "mu_0",
+    ],
+)
 def jax_calc_posterior_and_exceedances(
     theta_max,
     y,
     n,
     log_prior,
     neg_precQ,
```
Comment (tbenthompson) [inline research/berry/berrylib/fast_inla.py:401]: These two parameters are now unused because of the additions of eig_vals/vecs.

File: research/berry/berrylib/fast_inla.py [hunk 2]
```diff
@@ -328,59 +337,94 @@ def step(args):
...
+    else:
+        args = init_args
+        step = jax.jit(step)
+        for i in range(max_iter):
+            args = step(args)
+            out = args
+            if not args[-1]:
+                break
+    i, theta_max, hess_inv, go = out
     return theta_max, hess_inv
```
Comment (tbenthompson) [inline research/berry/berrylib/fast_inla.py:355]: Unless the performance cost is large, I think we should return `i` here too. It'll provide a nice degree of visibility into the optimizer and let us confirm that it converged from outside the jit jax code.

**Review by tbenthompson — APPROVED** [review]
This seems great. I went through the code and thought about stuff. I also pulled and ran it. I can confirm that it's much faster. No objections if you want to merge this unchanged, but I left a couple comments in case you want to make some small improvements. 

Two remaining questions:
1. How do I choose whether to run float32/float64? Or are we just assuming that we will only use float32 henceforth?
2. Where/how are you running benchmarks? The `if __name__=="__main__":` section at the end of `test_berry.py` is where I was putting a super simple benchmark before.

**Comment by constinit** [comment]
1) There are three options: 
- Selectively enable float64 in models that need it and run them in different Python processes 
- Enable float64 always, and have models cast back to float32 where they can
- Use a library to do the casting for us e.g. https://github.com/deepmind/jmp
2) Yes I was using your benchmark. I'm seeing a lot of noise on my system, but it's something like 20us inferences for baseline and 10us for this PR.

**Comment by tbenthompson** [comment]
Merge whenever you're ready!

> * Selectively enable float64 in models that need it and run them in different Python processes
> * Enable float64 always, and have models cast back to float32 where they can

Ok, so reading between the lines, there is currently no mechanism to choose between 64 and 32 bit float in the fast_inla.py code. It's just set to 32 with this PR? Is that correct?

**Comment by tbenthompson** [comment]
Other comments:
* I don't think enabling float64 when we import jax means anything about whether we're *required* to use float64 or not. 
* I don't know how the jax casting rules compare to numpy, but I'm assuming that if you perform some computation on two float64 arrays, you'll get a float64 array back and the same with float32. Is that wrong?
* So, if we have a jax_dtype parameter for the FastINLA constructor, then we can cast all the jax arrays to `jax_dtype` after they are created and then we can cast input data to `jax_dtype` in the inference calls. The result is that the user can easily choose 32/64. Does that sound right?

**Comment by constinit** [comment]
Ah, sadly numpy has a tendency to promote to float64:
>At the moment, JAX by default enforces single-precision numbers to mitigate the Numpy API’s tendency to aggressively promote operands to double. 

So we cannot generally control the output type by casting the inputs.

I confirmed that our code is indeed doing this somewhere. 🤔

**Comment by tbenthompson** [comment]
> So we cannot generally control the output type by casting the inputs.

I think this is only partially correct. If we control both all array inputs **and all scalar inputs** (particularly scalar integers), then we can control output type. The precise behavior of numpy is described in: https://numpy.org/doc/stable/reference/generated/numpy.result_type.html#numpy.result_type 

This explanation was also useful for me: https://stackoverflow.com/questions/45949263/numpy-casting-float32-to-float64

Don't feel like you need to work on this though! It's probably not the most fun type of work and it's definitely not urgent at all!
